### ISSUE TITLE
BUG: Custom masker offset is not working properly

### ISSUE THREAD

**Issue by adamwawrzynski** [shap/shap#3674]
### Issue Description

I wanted to perturb only selected parts of the text using `BertModel` based on this page in the documentation: https://shap.readthedocs.io/en/latest/example_notebooks/text_examples/sentiment_analysis/Using%20custom%20functions%20and%20tokenizers.html#Create-an-explainer. I replaced `for m in re.finditer(r"\W", s):` and the following code with handling list of objects with `start_index` and `end_index` properties.

Maybe I don't understand how `offset_mapping` should work. I assume that `input_ids` contain list of string that I wan't to mask out and analyze it's influence on model prediction, and `offset_mapping` is the list of tuples of the start and end indices of a given text in `input_ids` based on original text. In other words if I have text `i am ever feeling nostalgic about the fireplace i will know that it is still on the property` and I wan't to mask word `fireplace` then the `input _ids=["fireplace"]` and `offset_mapping=[(38, 47)]`.

### Minimal Reproducible Example

```python
import datasets
import numpy as np
import pandas as pd
import scipy as sp
import torch
import transformers

import shap

# load the emotion dataset
dataset = datasets.load_dataset("emotion", split="train")
data = pd.DataFrame({"text": dataset["text"], "emotion": dataset["label"]})

# load the model and tokenizer
tokenizer = transformers.AutoTokenizer.from_pretrained(
    "nateraw/bert-base-uncased-emotion", use_fast=True
)
model = transformers.AutoModelForSequenceClassification.from_pretrained(
    "nateraw/bert-base-uncased-emotion"
).cuda()
labels = sorted(model.config.label2id, key=model.config.label2id.get)


# this defines an explicit python function that takes a list of strings and outputs scores for each class
def f(x):
    tv = torch.tensor(
        [
            tokenizer.encode(v, padding="max_length", max_length=128, truncation=True)
            for v in x
        ]
    ).cuda()
    attention_mask = (tv != 0).type(torch.int64).cuda()
    outputs = model(tv, attention_mask=attention_mask)[0].detach().cpu().numpy()
    scores = (np.exp(outputs).T / np.exp(outputs).sum(-1)).T
    val = sp.special.logit(scores)
    return val

method = "keywords"

# build an explainer by passing a transformers tokenizer
if method == "transformers tokenizer":
    explainer = shap.Explainer(f, tokenizer, output_names=labels)

# build an explainer by explicitly creating a masker
elif method == "default masker":
    masker = shap.maskers.Text(r"\W")  # this will create a basic whitespace tokenizer
    explainer = shap.Explainer(f, masker, output_names=labels)

# build a fully custom tokenizer
elif method == "custom tokenizer":
    import re

    def custom_tokenizer(s, return_offsets_mapping=True):
        """Custom tokenizers conform to a subset of the transformers API."""
        pos = 0
        offset_ranges = []
        input_ids = []
        for m in re.finditer(r"\W", s):
            start, end = m.span(0)
            offset_ranges.append((pos, start))
            input_ids.append(s[pos:start])
            pos = end
        if pos != len(s):
            offset_ranges.append((pos, len(s)))
            input_ids.append(s[pos:])
        out = {}
        out["input_ids"] = input_ids
        if return_offsets_mapping:
            out["offset_mapping"] = offset_ranges
        return out

    masker = shap.maskers.Text(custom_tokenizer)
    explainer = shap.Explainer(f, masker, output_names=labels)

# keywords masking
elif method == "keywords":
    import re

    def custom_tokenizer(s, return_offsets_mapping=True):
        """Custom tokenizers conform to a subset of the transformers API."""
        if s == "":
            return {
                "input_ids": [],
                "offset_mapping": [],
            }

        print(f"Input:{[s]}")
        pos = 0
        offset_ranges = []
        input_ids = []

        keywords_dict = [
            {
                "text": "nostalgic",
                "start": 18,
                "end": 27,
            },
            {
                "text": "fireplace",
                "start": 38,
                "end": 47,
            },
            {
                "text": "property",
                "start": 84,
                "end": 92,
            },
        ]


        offset_ranges = []
        input_ids = []
        for key_term in keywords_dict:
            print(key_term)
            start_index = key_term["start"]
            end_index = key_term["end"]

            if start_index != -1:
                input_ids.append(s[start_index:end_index])
                offset_ranges.append((start_index, end_index))

        out =  {
            "input_ids": input_ids,
            "offset_mapping": offset_ranges,
        }
        print(out)
        return out

    masker = shap.maskers.Text(custom_tokenizer)
    explainer = shap.PermutationExplainer(f, masker, output_names=labels)


max_evals = 200
print([data["text"][3]])
# sample content: i am ever feeling nostalgic about the fireplace i will know that it is still on the property
shap_values = explainer([data["text"][3]], max_evals=max_evals)
print(shap_values.data[0])
```

### Traceback

```shell
/home/adamw/repositories/unified-predictions-v2/.venv/lib/python3.10/site-packages/datasets/load.py:1486: FutureWarning: The repository for emotion contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/emotion
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(
Input:['...']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['', '', ''], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
Input:['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['nostalgic', 'fireplace', 'property'], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
Input:['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['nostalgic', 'fireplace', 'property'], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
['nostalgic about the ' 'fireplace i will know that it is still on the '
 'property']
```

### Expected Behavior

I expect that as a result of the SHAP explainer, I will get attributions for the list of strings passed in `input_ids` and values in `shap_values.data[0]` will be extracted from the original text based on `offset_mapping` indices.

```/home/adamw/repositories/unified-predictions-v2/.venv/lib/python3.10/site-packages/datasets/load.py:1486: FutureWarning: The repository for emotion contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/emotion
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(
Input:['...']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['', '', ''], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
Input:['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['nostalgic', 'fireplace', 'property'], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
Input:['i am ever feeling nostalgic about the fireplace i will know that it is still on the property']
{'text': 'nostalgic', 'start': 18, 'end': 27}
{'text': 'fireplace', 'start': 38, 'end': 47}
{'text': 'property', 'start': 84, 'end': 92}
{'input_ids': ['nostalgic', 'fireplace', 'property'], 'offset_mapping': [(18, 27), (38, 47), (84, 92)]}
['nostalgic', 'fireplace', 'property']
```

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

0.45.1

**Comment by shuklapraveen1** [comment]
I took a look at `Text.token_segments()` and noticed that the offsets aren't used as direct token spans. Instead, each segment is built from the start of the current offset to the start of the next one.

For example, offsets like:

```python
[(18, 27), (38, 47), (84, 92)]
```

produce segments such as:

```text
"nostalgic about the "
"fireplace i will know that it is still on the "
"property"
```

which seems to match the behavior reported in this issue.

Before proposing a fix, could you clarify whether this is the intended behavior for `Text` maskers, or if custom tokenizers are expected to provide one-to-one token spans via `offset_mapping`?
